core/src/main/java/com/linecorp/armeria/server/docs/DocService.java

```diff
@@ -259,13 +237,130 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
         return unwrap().serve(ctx, req);
     }
 
+    static final class SpecificationLoader {
+        private final ExampleSupport exampleSupport;
+        private final DocServiceFilter filter;
+        private final NamedTypeInfoProvider namedTypeInfoProvider;
+        private final AtomicReference<List<ServiceConfig>> servicesRef =
+                new AtomicReference<>(Collections.emptyList());
+        private final Map<String, CompletableFuture<AggregatedHttpFile>> files = new ConcurrentHashMap<>();
+        private static final String versionsPath = "/versions.json";
+        private static final String specificationPath = "/specification.json";
+        private static final Set<String> targetPaths = ImmutableSet.of(versionsPath, specificationPath);
+
+        SpecificationLoader(
+                ExampleSupport exampleSupport,
+                DocServiceFilter filter,
+                @Nullable NamedTypeInfoProvider namedTypeInfoProvider) {
+            this.exampleSupport = exampleSupport;
+            this.filter = filter;
+            this.namedTypeInfoProvider = composeNamedTypeInfoProvider(namedTypeInfoProvider);
+        }
+
+        boolean contains(String path) {
+            return targetPaths.contains(path);
+        }
+
+        void triggerPreload(Executor executor) {
+            targetPaths.forEach(path -> load(path, executor));
+        }
```

Comment:
How about consolidating `triggerPreload` and `updateServices`? and it seems `AtomicReference` could be removed.
```java
void triggerPreload(List<ServiceConfig> services, Executor executor) {
    this.services = services;
    targetPaths.forEach(path -> load(path, executor));
}
```